components/wifi/esp8266-driver/ESP8266/ESP8266.cpp

```diff
@@ -446,27 +447,25 @@ int8_t ESP8266::rssi()
 
 int ESP8266::scan(WiFiAccessPoint *res, unsigned limit)
 {
-    unsigned cnt = 0;
-    nsapi_wifi_ap_t ap;
-
     _smutex.lock();
+
     set_timeout(ESP8266_CONNECT_TIMEOUT);
 
-    if (!_parser.send("AT+CWLAP")) {
-        _smutex.unlock();
-        return NSAPI_ERROR_DEVICE_ERROR;
-    }
+    _scan_r.res = res;
+    _scan_r.limit = limit;
+    _scan_r.cnt = 0;
 
-    while (_recv_ap(&ap)) {
-        if (cnt < limit) {
-            res[cnt] = WiFiAccessPoint(ap);
-        }
-
-        cnt++;
-        if (limit != 0 && cnt >= limit) {
-            break;
+    if (_parser.send("AT+CWLAP") && _parser.recv("OK\n")) {
```

Comment:
This is not correct. Should be
`if (!_parser.send("AT+CWLAP") || !_parser.recv("OK\n")) {` 
instead

Reply:
https://en.wikipedia.org/wiki/De_Morgan%27s_laws

`!(A && B) == !A || !B`

Reply:
Fixed